test/python/circuit/test_control_flow_builders.py

```diff
@@ -0,0 +1,2191 @@
+
+            expected = QuantumCircuit(qubits, clbits)
+            expected.for_loop(None, range(2), body, qubits, [clbits[0]])
+
+            self.assertCircuitsEquivalent(test, expected)
+
+        with self.subTest("while"):
+            cond = (clbits[0], 0)
+            test = QuantumCircuit(qubits, clbits)
+            with test.while_loop(cond):
+                test.break_loop()
+                test.h(0)
+                test.continue_loop()
+                test.measure(1, 0)
+
+            body = QuantumCircuit(qubits, [clbits[0]])
+            body.break_loop()
+            body.h(qubits[0])
+            body.continue_loop()
+            body.measure(qubits[1], clbits[0])
+
+            expected = QuantumCircuit(qubits, clbits)
+            expected.while_loop(cond, body, qubits, [clbits[0]])
+
+            self.assertCircuitsEquivalent(test, expected)
+
+    @ddt.data(QuantumCircuit.break_loop, QuantumCircuit.continue_loop)
+    def test_break_continue_accept_c_if(self, loop_operation):
+        """Test that ``break`` and ``continue`` statements accept :meth:`.Instruction.c_if` calls,
+        and that these propagate through correctly."""
+        qubits = [Qubit(), Qubit()]
+        clbits = [Clbit(), Clbit()]
+
+        with self.subTest("for"):
+            test = QuantumCircuit(qubits, clbits)
+            with test.for_loop(None, range(2)):
+                test.h(0)
+                loop_operation(test).c_if(1, 0)
+
+            body = QuantumCircuit([qubits[0]], [clbits[1]])
+            body.h(qubits[0])
+            loop_operation(body).c_if(clbits[1], 0)
+
+            expected = QuantumCircuit(qubits, clbits)
+            expected.for_loop(None, range(2), body, [qubits[0]], [clbits[1]])
+
+            self.assertCircuitsEquivalent(test, expected)
+
+        with self.subTest("while"):
+            cond = (clbits[0], 0)
+            test = QuantumCircuit(qubits, clbits)
+            with test.while_loop(cond):
+                test.h(0)
+                loop_operation(test).c_if(1, 0)
+
+            body = QuantumCircuit([qubits[0]], clbits)
+            body.h(qubits[0])
+            loop_operation(body).c_if(clbits[1], 0)
+
+            expected = QuantumCircuit(qubits, clbits)
+            expected.while_loop(cond, body, [qubits[0]], clbits)
+
+            self.assertCircuitsEquivalent(test, expected)
+
+    @ddt.data(QuantumCircuit.break_loop, QuantumCircuit.continue_loop)
+    def test_break_continue_only_expand_to_nearest_loop(self, loop_operation):
+        """Test that a ``break`` or ``continue`` nested in more than loop only expands as far as the
+        inner loop scope, not further."""
+        qubits = [Qubit(), Qubit(), Qubit()]
+        clbits = [Clbit(), Clbit(), Clbit()]
+        cond_inner = (clbits[0], 0)
+        cond_outer = (clbits[1], 0)
+
+        with self.subTest("for for"):
+            test = QuantumCircuit(qubits, clbits)
+            with test.for_loop(None, range(2)):
```

Comment:
This is a question for Kevin @kdk, I think.

Reply:
Good question. We picked `(loop_parameter, indexset, body)` initially to match the written order of `for i in {1, 3 5} { ... }`, though I agree it does make the case where you want a throwaway parameter via `None` considerably more verbose.

The alternative I suppose would be an ordering like `(indexset, body, loop_parameter=None)` (where `loop_parameter` would have to come last as an optional arg) which seems less intuitive in the common case to me, but I'd be curious to other opinions.

Reply:
At least from my perspective, I would think we should do whatever is the most natural for the builder interface, since the usability of these methods when specifying an explicit body parameter already isn't great (for me, it's difficult to reason about scoping without inline visual blocks).

>  it does make the case where you want a throwaway parameter via None considerably more verbose.

Agreed. The other thing is, we're not really saying we don't want a parameter by specifying `None`. We're saying we want one, but we want it generated, and to go away if we never use it.

I like the last ordering personally `(indexset, body, loop_parameter=None)`, since it I would _imagine_ it'd be uncommon to specify a parameter if using the builder interface, and hence leaving `body` unspecified / `None`. Is this the case? Can you clarify what makes this ordering not so intuitive?

Reply:
@kdk: in the full ordering, you have `qubits` and `clbits` as required arguments as well, which is why we originally just kept the loop parameter at the front.  I'd tend to agree with Kevin H that we probably want to optimise the ergonomics for the builder interface rather than the full form, because I think it'll be how more people use them.

For the builders, the argument of `for x in indexset` doesn't quite hold the same, because they look like
```python
with qc.for_loop(None, indexset) as x:
    pass
```
so it's more like `indexset -> x`, even now when we're passing the loop parameter as the first variable.

Reply:
Universally changed the order across all `for`-loop components in 1287a0461.